File: 4.cpp
#include <iostream>
#include <vector>

using namespace std;

double findMedianSortedArrays(vector<int>& nums1, vector<int>& nums2) {
    int m = nums1.size();
    int n = nums2.size();
    double ret;
    return ret;
}

void test() {
    vector<int> nums1;
    vector<int> nums2;
    double ret;

    cout << "nums1:{1, 3}, nums2:{2}" << endl;
    nums1 = {1, 3};
    nums2 = {2};
    ret = findMedianSortedArrays(nums1, nums2);
    cout << "ret:" << ret << endl;

    cout << "nums1:{1, 2}, nums2:{3, 4}" << endl;
    nums1 = {1, 2};
    nums2 = {3, 4};
    ret = findMedianSortedArrays(nums1, nums2);
    cout << "ret:" << ret << endl;
}

int main()
{
    test();
    return 0;
}

File: 4.py
def findMediansortedArrays(nums1, nums2):
    if len(nums1) > len(nums2):
        nums1, nums2 = nums2, nums1
    x, y = len(nums1), len(nums2)
    low, high = 0, x

    while low <= high:
        partitionX = (low + high) // 2
        partitionY = (x + y + 1) // 2 - partitionX

        maxX = float('-inf') if partitionX == 0 else nums1[partitionX - 1]
        minX = float('inf') if partitionX == x else nums1[partitionX]

        maxY = float('-inf') if partitionY == 0 else nums2[partitionY - 1]
        minY = float('inf') if partitionY == y else nums2[partitionY]

        if maxX <= minY and maxY <= minX:
            if (x + y) % 2 == 0:
                return (max(maxX, maxY) + min(minX, minY)) / 2.0
            else:
                return max(maxX, maxY)
        elif maxX > minY:
            high = partitionX - 1
        else:
            low = partitionX + 1


def test_findMediansortedArrays():
    nums1 = [1, 3]
    nums2 = [2]
    assert findMediansortedArrays(nums1, nums2) == 2.0

    nums1 = [1, 2]
    nums2 = [3, 4]
    assert findMediansortedArrays(nums1, nums2) == 2.5

    print("test passed")


test_findMediansortedArrays()
